### PR TITLE
Habit-Tracker-JVR-Hannes

### PR DESCRIPTION
**Pull request by JvR-Hannes**

### CODE REVIEW

**Comment by chrisjamiecarter** [comment]
Hey @JvR-Hannes 👋 ,

Just a quick one before I begin looking at your code.

When you submit a project, remember to submit the correct url
 - https://github.com/TheCSharpAcademy/CodeReviews.Console.HabitTracker/pull/199

When viewing the pull request from the Academy, it links to the project repo 😄 
- https://github.com/TheCSharpAcademy/CodeReviews.Console.HabitTracker

So I had to do some detective work to link them 🕵️ ...

Thanks
@chrisjamiecarter

**Review by chrisjamiecarter — CHANGES_REQUESTED** [review]
Hey @JvR-Hannes,

Well done on another project submission 🎉 

Unfortunately, you have forgotten to meet one [requirement](https://thecsharpacademy.com/project/12/habit-logger).

- [x] Your project needs to contain a Read Me file where you'll explain how your app works.

Please add a README file as per requirements 😄 .

Thanks
@chrisjamiecarter

**Comment by JvR-Hannes** [comment]
Hi Chris,

My sincerest apologies about that. I felt like somewhere I missed something
or missed a step.

Thank you for informing me, I will look into it and try not to do it in the
future again.

Thanks you Chris.

Regards

**Comment by JvR-Hannes** [comment]
Hi Chris

Is it possible to change the name locally and push it or will it affect my
app?

Thanks

**Comment by chrisjamiecarter** [comment]
@JvR-Hannes Yes make you changes locally, commit, then push to update this PR.

Any issues let me know and I will be happy to support.
Thanks
@chrisjamiecarter

**Comment by JvR-Hannes** [comment]
@chrisjamiecarter I have changed the name as you requested. I'm still struggling a bit with the name changes afterwards and the pull requests etc.
I will do some more homework to improve on this matter and hope that in the future, it is either less or no mistakes like this.

Thank you @chrisjamiecarter

File: Habit-Tracker-JVR-Hannes/Habit-Tracker/Program.cs [hunk 1]
```diff
@@ -0,0 +1,346 @@
...
+
+            if (recordId == 0)
+                return;
+
+            using (var connection = new SqliteConnection(connectionString))
+            {
+                connection.Open();
+
+                var checkCmd = connection.CreateCommand();
+                checkCmd.CommandText = $"SELECT EXISTS(SELECT 1 FROM habits WHERE Id = {recordId})";
```
Comment (chrisjamiecarter) [inline Habit-Tracker-JVR-Hannes/Habit-Tracker/Program.cs:151]: Would love to see you parameterise all of your queries like you have done elsewhere in your code.

Although, the ones I have seen are fine as you are validating user input and dealing with `int` and `DateTime`.

File: Habit-Tracker-JVR-Hannes/Habit-Tracker/Program.cs [hunk 2]
```diff
@@ -0,0 +1,346 @@
...
+                tableCmd.CommandText = $"UPDATE habits SET date = @date, quantity = @quantity WHERE Id = @recordId";
+                tableCmd.Parameters.AddWithValue("@date", date);
+                tableCmd.Parameters.AddWithValue("@quantity", quantity);
+                tableCmd.Parameters.AddWithValue("@recordId", recordId);
+
+                tableCmd.ExecuteNonQuery();
+
+                connection.Close();
+            }
+
```
Comment (chrisjamiecarter) [inline Habit-Tracker-JVR-Hannes/Habit-Tracker/Program.cs:175]: Remember you Spacing [code-conventions](https://thecsharpacademy.com/article/58/code-conventions)

- Remove empty lines after opening curly brackets and before closing curly brackets.

image

File: Habit-Tracker-JVR-Hannes/Habit-Tracker/Properties/habit-Tracker.db
Comment (chrisjamiecarter) [inline Habit-Tracker-JVR-Hannes/Habit-Tracker/Properties/habit-Tracker.db:1]: Ideally databases should not be stored in source control.

File: Habit-Tracker-JVR-Hannes/Habit-Tracker/Properties/launchSettings.json [hunk 3]
```diff
@@ -0,0 +1,8 @@
+{
+  "profiles": {
+    "Habit-Tracker": {
+      "commandName": "Project",
+      "workingDirectory": "C:\\Users\\Hannes\\source\\repos\\Habit-Tracker\\Habit-Tracker\\Properties\\"
```
Comment (chrisjamiecarter) [inline Habit-Tracker-JVR-Hannes/Habit-Tracker/Properties/launchSettings.json:5]: Be careful with hardcoding workingDirectory paths.

This path does not exist on my machine, so throws an exception on start-up.

If you are relying on this, add a useful not in your README under a Installation/Getting Started/Usage section.

**Review by chrisjamiecarter — APPROVED** [review]
Hey @JvR-Hannes 👋 ,

Fantastic job on your another project submission 🎉

Thank you for persisting with correcting some mistakes early on! It makes my life as a reviewer easier 😄 

I have added some minor comments that I feel can quickly support your growth as a developer.

And I have a few suggestions that I believe could help you enhance this application even further:

- You could clear the console each time before you write the menu. This would allow the user to segment an action, maintain readability of the console window and improve the experience for the user.
- If you present a list of information, like the habits, implement a `Console.WriteLine("Press any key to continue...");` and a `Console.Read();` after you have displayed the data. This will allow the user to read and digest the list before going back to a clean main menu (previous point).

**Bugs**
Unfortunately, I did spot a bug within your application:

As you can see, I have records in the database:
image

But when I View by Date Range, I get no records found:
image

I suggest revisiting to try understand and fix any issues, but as these relate to the extra challenge requirement they will not block the PR.

Feel free to reach out if you have any questions or if you'd like to collaborate further on any of these points.

I will go ahead and mark as approved, keep up the excellent work on the next projects! 😊

Best regards,
@chrisjamiecarter

**Pull request closed without merging**